cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/DefaultMetricDatasetFactory.java

```diff
@@ -114,30 +123,23 @@ public MetricsConsumerMetaTable createConsumerMeta() {
     }
   }
 
-  private MetricsTable getOrCreateMetricsTable(String tableName, DatasetProperties props) {
-    MetricsTable table = null;
+  private MetricsTable getOrCreateMetricsTable(String tableName, final DatasetProperties props) {
     // metrics tables are in the system namespace
-    DatasetId metricsDatasetInstanceId = NamespaceId.SYSTEM.dataset(tableName);
-    while (table == null) {
-      try {
-        table = DatasetsUtil.getOrCreateDataset(dsFramework, metricsDatasetInstanceId,
-                                                MetricsTable.class.getName(), props, null, null);
-      } catch (DatasetManagementException | ServiceUnavailableException e) {
-        // dataset service may be not up yet
-        // todo: seems like this logic applies everywhere, so should we move it to DatasetsUtil?
-        LOG.warn("Cannot access or create table {}, will retry in 1 sec.", tableName);
-        try {
-          TimeUnit.SECONDS.sleep(1);
-        } catch (InterruptedException ie) {
-          Thread.currentThread().interrupt();
-          break;
+    final DatasetId metricsDatasetInstanceId = NamespaceId.SYSTEM.dataset(tableName);
+    MetricsTable table;
+    try {
+      table = Retries.callWithRetries(new Retries.Callable<MetricsTable, Exception>() {
+        @Override
+        public MetricsTable call() throws Exception {
+          return DatasetsUtil.getOrCreateDataset(dsFramework, metricsDatasetInstanceId, MetricsTable.class.getName(),
+                                                 props, null, null);
         }
-      } catch (IOException e) {
-        LOG.error("Exception while creating table {}.", tableName, e);
-        throw Throwables.propagate(e);
-      }
+      }, RetryStrategies.fixDelay(1, TimeUnit.SECONDS), RETRYABLE_PREDICATE);
```

Comment:
For failing in getting dataset, I think it's better to have caller to decide what to do. E.g. for the operation stat case, it can simply just catch the service unavailable exception (assuming we propagate properly) and ignore it (without logging) as it'll get retried in the next interaction anyway.

The same code path is used by the metrics system. For the metrics processor, it might want to keep retrying until it succeed as we don't want losing any metrics. It can be done as part of the polling logic in the metrics processor when it is polling from TMS (simply just don't advance the offset if the writing was not successful). This means retry shouldn't be done in this method.

For the metrics handler, as mentioned above, retry is actually not needed (or only a very short retry).